---

# (pt-br) Projeto Guiado 2: Explorando postagens no Hacker News 
>Dataquest - Trilha Analista de Dados em Python - [Link](https://www.dataquest.io/path/data-analyst/)

# (us-en) Guided Project 2: Exploring Hacker News Posts
>Dataquest - Data Analyst in Python Path - [Link](https://www.dataquest.io/path/data-analyst/)
---
## Sobre o projeto:

Neste projeto, trabalharei com um conjunto de dados de submissões para o site de tecnologia popular Hacker News.
O Hacker News é um site iniciado pela incubadora de inicialização Y Combinator, onde as histórias enviadas pelos usuários (conhecidas como "posts") são votadas e comentadas, semelhante ao reddit. O Hacker News é extremamente popular nos círculos de tecnologia e startup, e as postagens que chegam ao topo das listagens da Hacker News podem obter centenas de milhares de visitantes como resultado.

## Objetivo do projeto:

O objetivo é demonstrar através da análise exploratória, a comparação entre dois tipos de postagens que são questionamentos e exibição conhecidos como "Ask HN" e "Show HN" verificando assim suas nunces.

## Abordagem do projeto:

**1. Interesses;**

Estamos especificamente interessados em publicações cujos títulos começam com Ask HN ou Show HN. Os usuários enviam postagens do Ask HN para fazer uma pergunta específica à comunidade do Hacker News. Abaixo estão alguns exemplos:

* Pergunte HN: Como melhorar meu site pessoal?
* Pergunte HN: Eu sou o único indignado com o Twitter fechando as contagens de compartilhamento?
* Perguntar HN: Aby alterações recentes no CSS que quebrou o celular?

Da mesma forma, os usuários enviam postagens do Show HN para mostrar à comunidade do Hacker News um projeto, produto ou apenas algo geralmente interessante. Abaixo estão alguns exemplos:

* Show HN: Wio Link Plataforma de Desenvolvimento de Hardware Web of Things baseada em ESP8266 '
* Mostre HN: Algo sem sentido que eu fiz
* Show HN: Shanhu.io, um playground de programação desenvolvido pela e8vm
    
**2. Processos de ETL (Extração, Transformação, Carregamento);**

   - Conjunto de dados: "hacker_news.csv"
   - Verificação de valores "missing", duplicados e etc.

**3. Análise exploratória;**

Vamos comparar esses dois tipos de postagens para determinar o seguinte:

Ask HN ou Show HN recebem mais comentários, em média? As postagens criadas em determinado horário recebem mais comentários, em média?

---
## About the project:

In this project, I'll work with a data set of submissions to popular technology site Hacker News.
Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

## Project's goal:

The objective is to demonstrate through the exploratory analysis, the comparison between two types of postings that are questioning and exhibition known as "Ask HN" and "Show HN" thus verifying their nunces.

## Project Approach:

**1. Interests;**

We're specifically interested in posts whose titles begin with either Ask HN or Show HN. Users submit Ask HN posts to ask the Hacker News community a specific question. Below are a couple examples:

* Ask HN: How to improve my personal website?
* Ask HN: Am I the only one outraged by Twitter shutting down share counts?
* Ask HN: Aby recent changes to CSS that broke mobile?

Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting. Below are a couple of examples:

* Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform'
* Show HN: Something pointless I made
* Show HN: Shanhu.io, a programming playground powered by e8vm

**2. Processes of ETL (Extraction, Transformation, Loading);**

   - Data set: "hacker_news.csv"
   - Check missing values, duplicate and etc.

**3. Exploratory analysis;**

We'll compare these two types of posts to determine the following:

Do Ask HN or Show HN receive more comments on average?
Do posts created at a certain time receive more comments on average?

---
### (pt-br) Começando a contar histórias

Comece abrindo e explorando o conjunto de dados 'open_dataset()'. Para facilitar, chame a função denominada 'explore_data ()' que pode ser usada repetidamente para imprimir linhas de maneira mais legível.

### (en-us) Beginning the storytelling

Start by opening and exploring these two data sets 'open_dataset()'. To make it easier, call the function named 'explore_data()' that it can use repeatedly to print rows in a more readable way.

In [1]:
def open_dataset(file_name):
    from csv import reader
    opened_file = open(file_name, encoding='utf8')
    read_file = reader(opened_file)
    data = list(read_file)
    return data

hn = open_dataset('hacker_news.csv')

def explore_data(dataset, start, end, rows_and_columns=False): # Selecione 'True' para imprimir a contagem de linhas e colunas.  Select True to print the count for rows and columns.
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') 
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

(pt-br)

Abaixo, explorando o data set utilizando a função 'explore_data()' é obtido a visualização do cabeçalho, os dados e quantidade de linhas e coluna.

(en-us)

Below, exploring the dataset using the 'explore_data ()' function gets the header, data, and row and column count.

In [2]:
print(explore_data(hn,0,5,True)) # Somente o cabeçalho e as quatro primeiras linhas de dados.
                                 # Only the header and the first four rows of data.


['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52']


['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30']


['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20']


['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']


Number of rows: 20101
Number of columns: 7
None


---
### (pt-br) Removendo cabeçalho da Lista de Listas
### (us-en) Removing Headers from a List of Lists


In [3]:
# Removendo o cabeçalho / # Removing headers
headers = hn[0]
hn = hn[1:]
print(headers)
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


(pt-br) Como podemos observar acima, para cada postagem temos um número X de comentários. Como estamos preocupados apenas com os títulos dos posts iniciados por "Ask HN" ou "Show HN", criaremos novas listas de listas contendo apenas os dados desses títulos.

(us-en) As we can see above, for each post we have an X number of comments. As we are only concerned with the titles of the posts initiated by "Ask HN" or "Show HN", we will create new lists of lists containing only the data of these titles.



---
### (pt-br) Separando postagens "Ask HN" e "Show HN"
### (us-en) Separating posts "Ask HN" and "Show HN"

In [4]:
ask_posts = []
show_posts = []
other_posts = []

for post in hn:
    title = post[1]
    if title.lower().startswith("ask hn"): #verificando se a postagem em "minúsuclo (lower)" e se "começa com (startswith)" ask hn
        ask_posts.append(post)
    elif title.lower().startswith("show hn"): #verificando se a postagem em "minúsuclo (lower)" e se "começa com (startswith)" show hn
        show_posts.append(post)
    else:
        other_posts.append(post) # postagem restantes armazenando na lista other_posts

print("Ask HN posts:",len(ask_posts))
print("Show HN posts:",len(show_posts))
print("Other HN posts:",len(other_posts))     

Ask HN posts: 1744
Show HN posts: 1162
Other HN posts: 17194


---
### (pt-br) Determinando se ask posts ou show posts recebem mais comentários em média.
### (us-en) Determining if ask posts or show posts receive more comments on average.

In [5]:
# Média de Ask Posts
# Average of Ask Posts

total_ask_comments = 0

for ask in ask_posts:
    num_comments = ask[4]
    total_ask_comments += int(num_comments)

avg_ask_comments = total_ask_comments / len(ask_posts)
print("Média/Average Ask Posts:", avg_ask_comments)

#--------------------------------------------------------------------------

# Média de Show Posts
# Average of Show Posts

total_show_comments = 0

for show in show_posts:
    num_comments = show[4]
    total_show_comments += int(num_comments)

avg_show_comments = total_show_comments / len(show_posts)
print("Média/Average Show Posts:", avg_show_comments)

Média/Average Ask Posts: 14.038417431192661
Média/Average Show Posts: 10.31669535283993


**(pt-br) Show posts ou ask posts recebem mais comentários, em média?**
Conforme análise exploratória dos dados, em média as postagens do tipo perguntas faz com que os usuário tenham  mais interação recebendo assim mais comentários no Hacker News. Com esse resultado, a análise será concentrada nesse tipo de postagem.

**(us-en) Do show posts or ask posts receive more comments on average?**
According to an exploratory analysis of the data, on average the question-type posts make the users have more interaction thus receiving more comments in Hacker News. With this result, the analysis will be concentrated on this type of post.

---
### (pt-br) Verificando ask posts criados em determinado horário
Em seguida, será determinado se as postagens solicitadas em determinado horário têm maior probabilidade de atrair comentários. Será utilizado os seguintes passos para realizar esta análise:

* Cálculo da quantidade de mensagens solicitadas criadas em cada hora do dia, juntamente com o número de comentários recebidos.
* Cálculo do número médio de comentários que as postagens recebem por hora criada.

### (us-en) Checking ask posts created at certain time
It will then determine whether requested posts at a particular time are more likely to attract comments. The following steps will be used to perform this analysis:

* Calculate the amount of requested messages created at each time of day, along with the number of comments received.
* Calculate the average number of comments that posts receive per hour you create.


In [6]:
import datetime as dt # Importando o módulo datetime sob o alias "dt" / Importing the datetime module under the alias "dt"

In [7]:
# Calculando a quantidade de ask posts criadas durante cada hora do dia e o número de comentários recebidos.
# Calculating the amount of ask posts created during each hour of day and the number of comments received.

result_list = []

for post in ask_posts:
    result_list.append([post[6], int(post[4])])

counts_by_hour = {}
comments_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for row in result_list:
    # Usando o método datetime.strptime () para analisar a data e criar um objeto datetime, depois utilizo o método datetime.strftime () para selecionar apenas a hora do objeto datetime.
    # Using the datetime.strptime () method to parse the date and create a datetime object, then I use the datetime.strftime () method to select only the time of the datetime object.
    hour = dt.datetime.strptime(row[0], date_format).strftime("%H")
    
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = row[1]
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += row[1]
comments_by_hour   

{'09': 251,
 '13': 1253,
 '10': 793,
 '14': 1416,
 '16': 1814,
 '23': 543,
 '12': 687,
 '17': 1146,
 '15': 4477,
 '21': 1745,
 '20': 1722,
 '02': 1381,
 '18': 1439,
 '03': 421,
 '05': 464,
 '19': 1188,
 '01': 683,
 '22': 479,
 '08': 492,
 '04': 337,
 '00': 447,
 '06': 397,
 '07': 267,
 '11': 641}

---
### (pt-br) Cálculo da média de comentários por hora
Em seguida, será usado os dois dicionários para calcular o número médio de comentários para postagens criadas durante cada hora do dia.
### (us-en) Calculation of the average comments by hour
The two dictionaries will then be used to calculate the average number of comments for posts created during each hour of the day.

In [8]:
avg_by_hour = []

for comment in comments_by_hour:
    avg_by_hour.append([comment, comments_by_hour[comment] / counts_by_hour[comment]])

avg_by_hour

[['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034]]

---
### (pt-br) Ordenando a lista de listas da média de comentários por hora
Para finalizar, classificando a lista de listas e imprimindo os cinco valores mais altos em um formato mais fácil de ler.
### (us-en) Sorting the list of lists of the average comments by hour
Finally, sorting the list of lists and printing the five highest values ​​in an easier format to read.

In [9]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1],row[0]]) #fazendo a troca "swap" entre comentário/hora
print(swap_avg_by_hour)

print('\n')

sorted_swap = sorted(swap_avg_by_hour, reverse=True) #classificando da maior média para a menor.
sorted_swap 

[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]




[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [10]:
print("Top 5 Horas para Comentários 'Ask HN':")
print("Top 5 Hours for 'Ask HN' Comments:")
for row, hour in sorted_swap[:5]:
    print("{}: {:.2f} Média de cometários por post / average comments per post".format(
        dt.datetime.strptime(hour,"%H").strftime("%H:%M"),row)
     )    

Top 5 Horas para Comentários 'Ask HN':
Top 5 Hours for 'Ask HN' Comments:
15:00: 38.59 Média de cometários por post / average comments per post
02:00: 23.81 Média de cometários por post / average comments per post
20:00: 21.52 Média de cometários por post / average comments per post
16:00: 16.80 Média de cometários por post / average comments per post
21:00: 16.01 Média de cometários por post / average comments per post


(pt-br) A hora que recebe mais comentários por postagem em média é 15:00, com uma média de 38,59 comentários por postagem. Há um aumento de 60% no número de comentários entre as horas com o maior e o segundo maior número médio de comentários.
De acordo com a documentação do conjunto de dados, o fuso horário utilizado é a Eastern Time nos EUA. Então, nós também podemos escrever 15:00 como 15:00 est. Com relação ao horário de Brasília (Brasil) adiantar em 2 horas, ou seja, é equivalente às 17:00 o período com mais postagens segundo o fuso horário brasileiro.

(us-en) The time you receive the most comments per post on average is 3:00 PM, with an average of 38.59 comments per post. There is a 60% increase in the number of comments between the hours with the highest and the second highest average number of comments.
According to the dataset documentation, the time zone used is Eastern Time in the USA. So, we can also write 3pm as 3pm EST. With respect to the Brasília time (Brazil) to advance in 2 hours, that is, the period with the most postings according to the Brazilian timezone is equivalent to 17:00.

---
### Conclusão / Fim da História
Neste projeto, foi analisado e mostrado as postagens para determinar qual tipo de postagem e horário recebem a maioria dos comentários em média. Com base na análise, para maximizar a quantidade de comentários que um post recebe, é recomendado que a postagem seja categorizada como ask post e criada entre 15:00 e 16:00 (3:00 pm est - 4:00 pm est).

No entanto, deve-se notar que o conjunto de dados analisado excluiu publicações sem comentários. Sendo assim, é mais correto dizer que, das postagens que receberam comentários, ask posts receberam mais comentários em média, e criadas entre 15:00 e 16:00 (3:00 pm est - 4:00 pm est) receberam a mensagem a maioria dos comentários.

### Conclusion / End of Story
In this project, we analyzed and shown the posts to determine which type of post and time receive the most comments on average. Based on the review, to maximize the amount of comments a post receives, it is recommended that the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the analyzed data set excluded publications without comment. So, it is more accurate to say that from posts that received comments, ask posts received more comments on average, and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the message most of the comments.